### ISSUE TITLE
Unexpected behavior of st_difference() in sf compared to PostGIS results

### ISSUE THREAD

**Issue by josesamos** [r-spatial/sf#2483]
Apologies if this issue has already been addressed and I missed it, and also if I am making any mistakes. I perform the operations in PostGIS and get the expected result, but not in sf.

```{r, echo=TRUE, eval=TRUE}
crs <- 4326
poly1 <- sf::st_polygon(
  list(rbind(c(0, 0), c(0, 4), c(4, 4), c(4, 0), c(0, 0)))
) |>
  sf::st_sfc(crs = crs)

line1 <- sf::st_linestring(rbind(c(2, -1), c(2, 5))) |>
  sf::st_sfc(crs = crs)
line2 <- sf::st_linestring(rbind(c(2, 0.5), c(2, 3.5))) |>
  sf::st_sfc(crs = crs)
line3 <- sf::st_linestring(rbind(c(-1, 4), c(5, 4))) |>
  sf::st_sfc(crs = crs)

point1 <- sf::st_point(c(2, 0)) |>
  sf::st_sfc(crs = crs)

# 1
sf::st_difference(line1, line2)

# 2
sf::st_difference(line3, poly1)

# 3
sf::st_difference(point1, poly1)

# 4
sf::st_difference(point1, line1)

```
In PostGIS, the results of operations 1 and 2 are MULTILINESTRING((2 -1,2 0.5),(2 3.5,2 5)) and MULTILINESTRING((-1 4,0 4),(4 4,5 4)), respectively, while for 3 and 4 the result is POINT EMPTY. 

In sf: LINESTRING (2 -1, 2 5), LINESTRING (-1 4, 5 4), POINT (2 0) and POINT (2 0), respectively.

Thank you for your attention.

**Comment by tim-salabim** [comment]
sf uses S2 geometry when dealing with unprojected data, whereas postgis uses planar representation of the unprojected data. With `sf_use_s2(FALSE)` the results are the same as with postgis.

**Comment by josesamos** [comment]
Thank you very much for the information, I didn't know that!

**Comment by josesamos** [comment]
One more thing, is it correct for sf to work this way? There are no warnings during execution. In PostGIS, if geographical coordinates and the Geography type are used (to account for the Earth's curvature), the ST_Difference function is not defined.

**Comment by tim-salabim** [comment]
When I load sf this message is shown:

```r
library(sf)
Linking to GEOS 3.12.1, GDAL 3.8.4, PROJ 9.3.1; sf_use_s2() is TRUE
```
and when I switch off S2 I get the following warning:
```r
although coordinates are longitude/latitude, st_difference assumes that they are planar
```

**Comment by josesamos** [comment]
Yes, I get the same messages. However, I’m referring to the fact that, when calling st_difference without having set sf::sf_use_s2(FALSE) beforehand, the result obtained is not correct. It might be appropriate to at least warn that this function may not return a correct result unless it is explicitly configured to disable the use of S2.

**Comment by edzer** [comment]
Correct as defined by whom? We're talking about geometry on the sphere; there may be tolerances that are by default set to zero that need to be set to a non-zero value before things start working (as is often the case for GEOS/planar geometry).

**Comment by josesamos** [comment]
I’m sorry, I probably didn’t express myself clearly. What I mean is that the result is not the "expected" one, at least in a simple example like the one I’m presenting.

**Comment by josesamos** [comment]
An even simpler example and the result obtained:

```{r, echo=TRUE, eval=TRUE}
crs <- 4326

line4 <- sf::st_linestring(rbind(c(0, -1), c(3, -1))) |>
  sf::st_sfc(crs = crs)

point2 <- sf::st_point(c(3, -1)) |>
  sf::st_sfc(crs = crs)

sf::st_difference(point2, line4)

```
## POINT (3 -1)

**Comment by rsbivand** [comment]
Topology on the plane and on the sphere are different. If you set `"EPSG:4326"`, you are asserting that algorithms appropriate for spherical coordinates must be used. So planar outcomes are undefined in this setting. If you assert a planar CRS, algorithms appropriate for planar geometries must be used. You are asserting a spherical CRS but complain when a planar algorithm is not used. Recall that your line has a length of:
```
> sf_use_s2(TRUE)
Spherical geometry (s2) switched on
> st_length(line4)
333534.5 [m]
```
but planar:
```
> line4 |> st_set_crs(NA) |> st_length()
[1] 3
```
Why are you asserting `"EPSG:4326"`?

**Comment by josesamos** [comment]
I had just generated a simple example to show the use of the functions in PostGIS and sf. I understand that the line considering the curvature of the Earth has a different length than if it is considered plane geometry. But I expected that the result of this last example that I have included would be POINT EMPTY since it is one of the two points that define the line, even considering the curvature of the Earth.
